File: typescript/packages/core/src/vfs/wandb/wandb.ts
import { BoundVFS } from '../bound.ts'
import { WANDB_IO } from '../../commands/builtin/wandb/io.ts'
import { WandbAccessor } from '../../accessor/wandb.ts'

import { WANDB_COMMANDS } from '../../commands/builtin/wandb/index.ts'
import type { RegisteredCommand } from '../../commands/config.ts'
import { redactWandbConfig } from '../../core/wandb/config.ts'
import type { WandbConfig, WandbConfigRedacted } from '../../core/wandb/config.ts'

import { WANDB_OPS } from '../../ops/wandb/index.ts'
import type { RegisteredOp } from '../../ops/registry.ts'

import type { VFS } from '../../vfs/base.ts'
import { WANDB_PROMPT } from '../../vfs/wandb/prompt.ts'
import { VFSName } from '../../types.ts'

export interface WandbVFSState {
  type: string
  config: WandbConfigRedacted
}

export class WandbVFS extends BoundVFS<WandbAccessor> implements VFS {
  readonly kind: string = VFSName.WANDB
  readonly prompt: string = WANDB_PROMPT
  readonly config: WandbConfig
  readonly accessor: WandbAccessor

  constructor(config: WandbConfig) {
    super(WANDB_IO)
    this.config = config
    this.accessor = new WandbAccessor(config)
  }

  commands(): readonly RegisteredCommand[] {
    return WANDB_COMMANDS
  }

  ops(): readonly RegisteredOp[] {
    return WANDB_OPS
  }

  override getState(): Promise<WandbVFSState> {
    return Promise.resolve({
      type: this.kind,
      config: redactWandbConfig(this.config),
    })
  }
}

File: python/mirage/vfs/wandb/wandb.py
from typing import Any

from mirage.accessor.wandb import WandbAccessor
from mirage.commands.builtin.wandb import COMMANDS
from mirage.commands.builtin.wandb.io import IO
from mirage.core.wandb.config import WandbConfig
from mirage.ops.wandb import OPS
from mirage.types import VFSName
from mirage.vfs.bound import BoundVFS
from mirage.vfs.wandb.prompt import PROMPT


class WandbVFS(BoundVFS):
    name: str = VFSName.WANDB
    PROMPT: str = PROMPT

    def __init__(self, config: WandbConfig) -> None:
        super().__init__(io=IO)
        self.config = config
        self.accessor = WandbAccessor(config)
        for command in COMMANDS:
            self.register(command)
        for op in OPS:
            self.register_op(op)

    def get_state(self) -> dict[str, Any]:
        return self.config_state(self.config)
